includes/class-google-jwt.php

```diff
@@ -0,0 +1,175 @@
+
+	/**
+	 * The URL where we can find the Google OpenID configuration.
+	 *
+	 * @var string
+	 */
+	const GOOGLE_CONFIG_URL = 'https://accounts.google.com/.well-known/openid-configuration';
+
+	/**
+	 * The raw string with the received JWT token.
+	 *
+	 * @var string
+	 */
+	private $payload;
+
+	/**
+	 * Constructor
+	 *
+	 * @param string $payload The raw string with the received JWT token.
+	 */
+	public function __construct( $payload ) {
+		$this->payload = $payload;
+	}
+
+	/**
+	 * Get the JWKS URI from Google OpenID configuration.
+	 *
+	 * @return string|bool The JWKS URI or false if it could not be retrieved.
+	 */
+	public function get_jwks_uri() {
+		$response = wp_remote_get( self::GOOGLE_CONFIG_URL );
+		if ( is_wp_error( $response ) ) {
+			return false;
+		}
+
+		$body   = wp_remote_retrieve_body( $response );
+		$config = json_decode( $body, true );
+		if ( ! isset( $config['jwks_uri'] ) ) {
+			return false;
+		}
+
+		return $config['jwks_uri'];
+	}
+
+	/**
+	 * Get the JWKS from Google if the cache is expired and cache them.
+	 *
+	 * @return array|bool The JWKS or false if it could not be retrieved.
+	 */
+	public function get_jwks() {
+		if ( ! $this->should_refresh_jwks_cache() ) {
+			return $this->get_jwks_cached();
+		}
+
+		$jwks_uri = $this->get_jwks_uri();
+		if ( ! $jwks_uri ) {
+			return false;
+		}
+
+		$response = wp_remote_get( $jwks_uri );
+		if ( is_wp_error( $response ) ) {
+			return false;
+		}
+
+		$body = wp_remote_retrieve_body( $response );
+		$jwks = json_decode( $body, true );
+		if ( ! isset( $jwks['keys'] ) ) {
+			return false;
+		}
+
+		$this->update_jwks_cache( $jwks );
+
+		return $jwks;
+	}
+
+	/**
+	 * Update the JWKS cache.
+	 *
+	 * @param array $value The JWKS to cache.
+	 */
+	public function update_jwks_cache( $value ) {
+		update_option( self::CACHE_OPTION_NAME, $value );
+		update_option( self::CACHE_TIMESTAMP_OPTION_NAME, time() );
+	}
+
+	/**
+	 * Get the JWKS from cache.
+	 *
+	 * @return array|bool The JWKS or false if it could not be retrieved.
+	 */
+	public function get_jwks_cached() {
+		return get_option( self::CACHE_OPTION_NAME, false );
+	}
+
+	/**
+	 * Check if we should refresh the JWKS cache.
+	 *
+	 * Cache is refreshed anytime a signature verification fails, but only if it's older than 5 minutes to avoid abuse.
+	 */
+	public function should_refresh_jwks_cache() {
+		$timestamp = get_option( self::CACHE_TIMESTAMP_OPTION_NAME, 0 );
+		return time() - $timestamp > 300;
+	}
+
+	/**
+	 * Decode the JWT token.
+	 *
+	 * @return mixed|WP_Error The decoded token or a WP_Error if it could not be decoded.
+	 */
+	public function decode() {
+		$jwks = $this->get_jwks_cached();
+		if ( ! $jwks ) {
+			$jwks = $this->get_jwks();
+		}
+
+		if ( ! $jwks ) {
+			return false;
+		}
+		try {
+			$decoded = \Firebase\JWT\JWT::decode( $this->payload, \Firebase\JWT\JWK::parseKeySet( $jwks ) );
+		} catch ( SignatureInvalidException $e ) {
```

Comment:
```
[02-Jul-2024 23:35:32 UTC] PHP Fatal error:  Uncaught Firebase\JWT\SignatureInvalidException: Signature verification failed in /wordpress/plugins/newspack-plugin/4.4.0/vendor/firebase/php-jwt/src/JWT.php:150
Stack trace:
#0 /srv/htdocs/wp-content/plugins/newspack-extended-access/includes/class-google-jwt.php(160): Firebase\JWT\JWT::decode('eyJhbGciOiJSUzI...', Array)
#1 /srv/htdocs/wp-content/plugins/newspack-extended-access/includes/class-rest-controller.php(84): Newspack\ExtendedAccess\Google_Jwt->decode()
...
```
For some reason it's not catching here

Reply:
https://github.com/Automattic/newspack-extended-access/pull/5/commits/a5c0e673e32ea77c75532377e68eace8f3025d3e should fix it.

I don't know why but the `use` statements are not working as they should... see how I used the full namespace for the `JWT` class as well...

But did it happen on the staging site? How was it working for me then?